### PR TITLE
WL-3985 - Reimplement citations filepicker

### PR DESCRIPTION
**Pull request by surya-janani**
1.Redirect from citations helper to filepickerhelper using startHelper method and set mode to edit_attachment.
2.extractCitationFromRunData extracts citation from attribute FILE_PICKER_ATTACHMENTS in session.
3.On return from FilePickerHelper, buildEditAttachmentPanel makes use of the above method to extract citation,set in it state and pass it onto buildEditPanel.

Will remove commented code on top of other review changes that come in

### CODE REVIEW

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 1]
```diff
@@ -2223,12 +2244,36 @@ public String buildMainPanelContext(VelocityPortlet portlet, Context context, Ru
...
+			if(citation!=null) {
+				state.setAttribute(CitationHelper.CITATION_EDIT_ITEM, citation);
+				state.setAttribute(CitationHelper.CITATION_EDIT_ID,citation.getId());
+				//collection.add(citation);
+				state.setAttribute(STATE_CITATION_COLLECTION, collection);
+				buildEditPanelContext(portlet, context, data, state);
+			}
+
+		}catch(Exception e){
+
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: Need some error handling here, if it's expected that the user can trigger it just display to the user, if it's unexpected log it and display something to the user.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 2]
```diff
@@ -4059,17 +4138,24 @@ public void doReviseCitation (RunData data)
...
+						citation = collection.getCitation(citationId);
+					}
 		            String schemaId = params.getString("type");
 		            Schema schema = getCitationService().getSchema(schemaId);
 		            citation.setSchema(schema);
 
 		    		updateCitationFromParams(citation, params);
+					if(!collection.contains(citation)){
+						collection.add(citation);
+						getCitationService().save(collection);
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: Doesn't the collection get saved later in this method anyway?

**Comment by surya-janani** [comment]
Just needed the collection.add() . Removing the redundant save.

**Comment by buckett** [comment]
This is being put on hold as it clashes with the work Nick is doing.

**Comment by buckett** [comment]
Branch has conflicts that need resolving.

**Comment by buckett** [comment]
Also commented out code is still in there.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 3]
```diff
@@ -3003,7 +3048,7 @@ public void doFinish ( RunData data)
 			// Remove session collection
 			state.removeAttribute(STATE_CITATION_COLLECTION_ID);
 			state.removeAttribute(STATE_CITATION_COLLECTION);
-	
+
 			state.removeAttribute("fromListPage");
```
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:3077]: Indentation

Comment (surya-janani) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:3077]: Does indentation really matter on blank lines?

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 4]
```diff
@@ -829,9 +824,12 @@ public void doPost(HttpServletRequest req, HttpServletResponse res) throws Servl
 			} else if(requestedMimetype != null && requestedMimetype.equals(MIMETYPE_HTML)) {
 				doPostHtmlFragmentResponse(params, state, req, res);
 			}
-
+			/*if(params.get("hPickResource").equalsIgnoreCase("pickResource")) {
+				doPickResource(req,res);
+			}*/
 			return;
```
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:817]: We don't really need this commented out code here, do we?

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 5]
```diff
@@ -3003,7 +3035,7 @@ public void doFinish ( RunData data)
 			// Remove session collection
 			state.removeAttribute(STATE_CITATION_COLLECTION_ID);
 			state.removeAttribute(STATE_CITATION_COLLECTION);
-	
+		
 			state.removeAttribute("fromListPage");
```
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:3077]: It just makes the code harder to patch on Sakai as it's more likely that the line you've cleaned up will have been changed in Sakai and so cause a conflict when applying.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 6]
```diff
@@ -2223,12 +2231,36 @@ public String buildMainPanelContext(VelocityPortlet portlet, Context context, Ru
...
+		logger.debug("buildEditAttachmentPanelContext()");
+		ParameterParser params = data.getParameters();
+		state.removeAttribute("sakaiproject.filepicker.addAttachments");
+		try {
+			CitationCollection collection = getCitationCollection(state, true);
+			Citation citation = extractCitationFromRunData(data);
+			if(citation!=null) {
+				state.setAttribute(CitationHelper.CITATION_EDIT_ITEM, citation);
+				state.setAttribute(CitationHelper.CITATION_EDIT_ID,citation.getId());
+				//collection.add(citation);
```
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: commented out code

**Comment by surya-janani** [comment]
In order to resolve the conflicts, I've changed the exception Handling code in doReviseCitation method along with other changes.Please review it and let me know if it looks ok.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 7]
```diff
@@ -4072,42 +4138,36 @@ public void doReviseCitation (RunData data)
...
+				citation.setSchema(schema);
+				updateCitationFromParams(citation, params);
+				if(!collection.contains(citation)){
+					collection.add(citation);
+				}
+			}
+			// add citation to current collection
+			if (location==0){
+				getCitationService().save(collection);
 			}
```
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:4198]: Has anything other than indentation changed here?

Comment (surya-janani) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:4198]: Added  if(!collection.contains(citation)){
                    collection.add(citation);
            } 
and assigning citation in catch block

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 8]
```diff
@@ -560,7 +560,7 @@ citations_new_resource.init = function() {
 		return false;
 	});
 	$('#PickResource').on('click', function(eventObject) {
-		var target = $(eventObject.target);
+		/*var target = $(eventObject.target);
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: You shouldn't leave commented out code in unless there's a really good reason, if someone want it back they can look in the git history. If there is a good reason document it with a comment about the commented out section.

**Comment by surya-janani** [comment]
Removed the commented block

**Comment by buckett** [comment]
This still has some commented out code in the pull request.

**Comment by buckett** [comment]
Are we putting this one on hold? or is it going in?

**Comment by surya-janani** [comment]
I think I have to test this once with Nick's citation changes .

**Comment by surya-janani** [comment]
Fixing bits after merging with Nick's citation changes

**Comment by buckett** [comment]
Could you squash this down into one commit please.

**Comment by surya-janani** [comment]
I've rebased it

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 9]
```diff
@@ -3989,7 +4016,44 @@ public void doImportCitationFromResourceUrl( RunData data )
...
+		try {
+			SessionState session = ((JetspeedRunData)data).getPortletSessionState(((JetspeedRunData) data).getJs_peid());
+			if(session.getAttribute(FilePickerHelper.FILE_PICKER_ATTACHMENTS) != null) {
+				List refs = (List) session
+						.getAttribute(FilePickerHelper.FILE_PICKER_ATTACHMENTS);
+				String citationCollectionId = (String) sessionManager.getCurrentToolSession().getAttribute(STATE_CITATION_COLLECTION_ID);
+				CitationCollection collection = getCitationService().getCollection(citationCollectionId);
+				Citation citation = null;
+				if (!refs.isEmpty()) {
+					for (Iterator it = refs.iterator(); it.hasNext(); ) {
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: You iterate over all the references, but throw all the changes away apart from the last one, is there a reason for doing this?

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 10]
```diff
@@ -560,39 +560,13 @@ citations_new_resource.init = function() {
...
-					setTimeout(function() { citations_new_resource.watchForUpdates(jsObj.timestamp + 1); }
-						, citations_new_resource.secondsBetweenSaveciteRefreshes * 1000);
-				} catch (e) {
-					popUpError("Your browser has blocked a pop up. Please allow pop ups from WebLearn.");
-				}
+				$('#sakai_action').val('doPickResource');
+				$('#ajaxRequest').val('false');
+				$('#newCitationListForm').submit();
 			}
-		};
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js:595]: I think you should still have the trailing semicolon here as it's the end of the `var successObj  = {`

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java (hunk 9 above)
Comment (surya-janani) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: That list is always expected to have only one item ,becoz of file picker selection cardinality. I've removed the iteration and the unused collection variables

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 10 above)
Comment (surya-janani) [inline citations-tool/tool/src/webapp/js/new_resource.js:595]: I've added semicolon now, but it was working fine without semicolon

**Comment by surya-janani** [comment]
Made review changes

**Comment by buckett** [comment]
Could you just squash these into one commit please, especially as the first commit doesn't have a Jira reference.

**Comment by surya-janani** [comment]
rebased

**Pull request merged**